File: typescript/src/resources/geosignals.ts
/**
 * Official PIA SDK - Geopolitical Signals & Macro Conflict Map Resource
 */

import type { HttpTransport } from "../http/transport";
import type { RequestOptions } from "../types";

export interface GeoSignalEvent {
  id: string;
  title: string;
  region: string;
  severity: "low" | "medium" | "high" | "critical";
  category: "conflict" | "sanction" | "trade" | "election" | "supply_chain";
  latitude?: number;
  longitude?: number;
  impacted_assets?: string[];
  summary: string;
  published_at: string;
}

export interface GeoSignalsResponse {
  total: number;
  events?: GeoSignalEvent[];
  items?: GeoSignalEvent[];
}

export interface GeoSignalsMapResponse {
  layers: Array<{
    region: string;
    risk_level: number;
    active_conflicts: number;
    chokepoints_status?: Record<string, string>;
  }>;
}

export interface GeoSignalsAssetImpactResponse {
  assets: Array<{
    symbol: string;
    risk_score: number;
    primary_risk_driver: string;
    affected_supply_pct?: number;
  }>;
}

export class GeosignalsResource {
  constructor(private readonly transport: HttpTransport) {}

  /**
   * Retrieves active geopolitical risk events and conflict alerts.
   */
  public async getEvents(options?: RequestOptions): Promise<GeoSignalsResponse> {
    const raw = await this.transport.request<any>(
      "/api/v1/geosignals",
      "GET",
      undefined,
      options
    );
    return {
      ...raw,
      total: raw?.total ?? (raw?.items || []).length,
      items: raw?.items || raw?.events || [],
      events: raw?.events || raw?.items || [],
    };
  }

  /**
   * Retrieves global geopolitical risk map layer data and maritime chokepoints status.
   */
  public async getMap(options?: RequestOptions): Promise<GeoSignalsMapResponse> {
    return this.transport.request<GeoSignalsMapResponse>(
      "/api/v1/geosignals/map",
      "GET",
      undefined,
      options
    );
  }

  /**
   * Retrieves mapped asset vulnerabilities and commodity exposure to current geopolitical tensions.
   */
  public async getAssetImpacts(options?: RequestOptions): Promise<GeoSignalsAssetImpactResponse> {
    return this.transport.request<GeoSignalsAssetImpactResponse>(
      "/api/v1/geosignals/assets",
      "GET",
      undefined,
      options
    );
  }
}

File: python/pia/resources/geosignals.py
"""Official PIA SDK - Geopolitical Signals & Macro Conflict Map Resource."""

from __future__ import annotations

from typing import Any, Dict, Optional

from ..transport import AsyncTransport, SyncTransport


class GeosignalsResource:
    """Synchronous Geopolitical Signals & Risk Intelligence resource."""

    def __init__(self, transport: SyncTransport) -> None:
        self._transport = transport

    def get_events(self) -> Dict[str, Any]:
        """Retrieves active geopolitical risk events and conflict alerts."""
        return self._transport.request("/api/v1/geosignals", method="GET")

    def get_map(self) -> Dict[str, Any]:
        """Retrieves global geopolitical risk map layer data and maritime chokepoints."""
        return self._transport.request("/api/v1/geosignals/map", method="GET")

    def get_asset_impacts(self) -> Dict[str, Any]:
        """Retrieves mapped asset vulnerabilities to current geopolitical tensions."""
        return self._transport.request("/api/v1/geosignals/assets", method="GET")


class AsyncGeosignalsResource:
    """Asynchronous Geopolitical Signals & Risk Intelligence resource."""

    def __init__(self, transport: AsyncTransport) -> None:
        self._transport = transport

    async def get_events(self) -> Dict[str, Any]:
        """Asynchronously retrieves active geopolitical risk events."""
        return await self._transport.request("/api/v1/geosignals", method="GET")

    async def get_map(self) -> Dict[str, Any]:
        """Asynchronously retrieves global geopolitical risk map layer data."""
        return await self._transport.request("/api/v1/geosignals/map", method="GET")

    async def get_asset_impacts(self) -> Dict[str, Any]:
        """Asynchronously retrieves mapped asset vulnerabilities to geopolitical tensions."""
        return await self._transport.request("/api/v1/geosignals/assets", method="GET")
